pkg/services/ngalert/state/historian/annotation.go

```diff
@@ -0,0 +1,88 @@
+package historian
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/annotations"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/state"
+)
+
+// AnnotationStateHistorian is an implementation of state.Historian that uses Grafana Annotations as the backing datastore.
+type AnnotationStateHistorian struct {
+	annotations annotations.Repository
+	dashboards  dashboards.DashboardService
+	log         log.Logger
+}
+
+func NewAnnotationHistorian(annotations annotations.Repository, dashboards dashboards.DashboardService, log log.Logger) *AnnotationStateHistorian {
+	return &AnnotationStateHistorian{
+		annotations: annotations,
+		dashboards:  dashboards,
+		log:         log,
+	}
+}
+
+func (h *AnnotationStateHistorian) RecordState(ctx context.Context, rule *ngmodels.AlertRule, labels data.Labels, evaluatedAt time.Time, currentData, previousData state.InstanceStateAndReason) {
```

Comment:
Logically identical to `annotateState` from `manager.go`. Some vars have been renamed, that's it.

Reply:
I think this should be able to handle multiple states in a single call - too many small transactions is an ongoing performance problem. Giving the implementation the control over whether and how to batch writes is important.

I think it'd be ok to do this in a follow-up PR but I also think it's very important.

Reply:
Agree...eventually, but this is also meant to be a refactor.

The choice to make the args here identical to the old helper function was an intentional one. As a reviewer, I never like it when a chunk of code gets moved and the logic gets changed at the same time, it becomes hard to follow what exactly changed in the diff. So, I leaned towards changing the structure of moved code as little as possible within the same PR.

Reply:
I'd also like this to work in terms of DTOs more clearly - to use the actual `state.State` structs, and not the piecemeal fields that it takes right now. But, that takes some improvement of the code in the State package, which I felt was out of scope for this PR - that's what I'm looking into next.

I feel this is actually required before we can think about taking multiple states in the same call. The current structure isn't very conducive to it at all.

Reply:
I think if I had a better sense of what was coming next after this PR I'd be able to judge the direction better. Right now I'm concerned about emphasizing a difference between current state and history, and doing too much experimentation behind the Historian interface without fixing the state manager's bad practice of writing everything down twice.

Reply:
I dont feel that annotations are a stable enough backend to experiment with a solution to this problem just yet. The actual history implementation needs to iterate a bit, first.

Annotations, have their own eviction system that we do not control, many other things write to annotations, we have few guarantees about their durability.

There may be a point where we can do away with the instance table by calculating state based on the most recent transition in history. However, I don't think we'll be able to solve that well without a (rudimentary, even) state implementation that isn't based on annotations.

This is why I felt this abstraction layer needs to be added sooner rather than later. I want to try and reduce the coupling between these problems. I don't see this layer going away over the course of the project (maybe changing, sure).